text/0000-network-spec.md

```diff
@@ -68,42 +86,116 @@ When one peer creates outgoing connection to another peer, they send a `Handshak
+We expect that some nodes enter the network without being active validators (no staking). This nodes must be able to connect, and have access to fetch blockchain data.
+Since we expect that nodes keep their view of the network synced (all active connections between peers) we should avoid malicious nodes spamming honest nodes with large portions of useless connections that contains links to mostly ghost peer ids.
 
-We identify list of possible attacks, their severity and resulting reaction of the node:
+To address both of these issues we will be using *Reputation* and *Localization*.
 
-- Invalid signed handshake. Ban IP address. Note that user might end up in the situation when an attacker took over their router and via man-in-the-middle responds to messages with invalid signature. This node will be left without known peers, but we think this is the safest outcome and requires human intervention.
+### Reputation
 
-- Peer responds with false `latest_height` / `latest_weight`. The attacker might want to stall node in syncing mode, waiting for headers that don't exist by providing way to large `latest_height`. The only way to actually verify this is to try to pull headers from this peer. As we start requesting headers, a timer setup. If peer doesn't respond to the `HeaderRequest` within this time, we Ban the peer for a medium period of time. Additional heuristic can be added, which node will be banned if their `latest_height` is above some threshold of heights known from other peers.
+Reputation starting epoch `X` is a combination of:
+
+- Reputation starting epoch `X - 1`.
+- Stake on the network on epoch `X`.
+- Some sort of page rank algorithm on top of received endorsements.
+
+Reputation of one node can vary from different points of view. Reputation will be impacted from other factors such as how good does a node behave routing messages to and from other peers. This can be measured using ping/pong like messages. How much time have we been connected to this network. How relevant is information we are receiving from this peer.
+
+#### Ping/Pong
+
+To detect if a route from node `u` to node `w` through direct peer `v` exist we can send ping messages and expect pong message. This can be done on some timeout. This will allow to determine live routes, and can be used to increase/decrease reputation. Pong message will be sent back using routing back mechanism.
+
+### Localization
+
+We maintain a graph with all connections from our point of view. We should build a DAG/Tree (this should be similar to have several routes for some peer) in such a way that every path from us to other nodes is:
+
+- reasonably short.
+- use high reputation nodes.
+
+One idea to localize nodes in the network:
+
+- From our POV we have infinite reputation.
+- Build a directed graph where weight of edge [u -> v] is reputation of node [u].
```

Comment:
1. We will only use real edges on the network. The answer is yes [me -> other node] has infinite weight, but we will only use active connection.

2. Can we trust in what reputation `u` tells about `v`? In my mind we keep reputation as a local value, since peers might forge high reputation values of other peers. What you are telling makes sense, since that path up to `u` will be important/short if intermediates hops also have high reputation. We should take care of malicious actor directly (or indirectly) telling about other nodes (the same actor in disguise) having high reputation.